hail/src/main/scala/is/hail/Uploader.scala

```diff
@@ -0,0 +1,139 @@
+package is.hail
+
+import java.io.{PrintWriter, StringWriter}
+import java.net.InetAddress
+
+import is.hail.utils._
+import java.util.concurrent.LinkedBlockingQueue
+
+import is.hail.expr.ir.{BaseIR, Pretty}
+import org.apache.http.client.methods.HttpPost
+import org.apache.http.entity.StringEntity
+import org.apache.http.impl.client.HttpClientBuilder
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager
+import org.apache.http.util.EntityUtils
+import org.json4s.JsonAST.{JInt, JObject, JString}
+import org.json4s.jackson.JsonMethods
+
+import scala.collection.JavaConverters._
+
+object Uploader {
+  private lazy val theUploader: Uploader = new Uploader
+
+  var pipelineEmail: String = _
+
+  def uploadPipeline(ir0: BaseIR, ir: BaseIR) {
+    if (pipelineEmail == null)
+      return
+
+    // for stack trace
+    val e = new Traceback
+
+    val w = new StringWriter
+    // closes w
+    val stackTrace = using(new PrintWriter(w)) { p =>
+      e.printStackTrace(p)
+      w.toString
+    }
+
+    val contents =
+      s"ir0:\n${ Pretty(ir0) }\n\nir:\n${ Pretty(ir) }\n\nfrom:\n${ stackTrace }"
+
+    theUploader.enqueueUpload("ir", contents, pipelineEmail)
+  }
+
+  def upload(typ: String, contents: String, email: String) {
+    theUploader.upload(typ, contents, email)
+  }
+}
+
+class Uploader { self =>
+
+  private val config = {
+    val hc = HailContext.get
+    val sc = hc.sc
+    val hConf = hc.hadoopConf
+    val runtime = Runtime.getRuntime
+
+    JObject(
+      "hostname" -> JString(InetAddress.getLocalHost.getHostName),
+      "jvm_version" -> JString(System.getProperty("java.version")),
+      "jvm_properties" -> JObject(
+        System.getProperties.asScala.map { case (k, v) =>
+          k -> JString(v)
+        }
+          .toList),
+      "jvm_runtime" -> JObject(
+        "available_processors" -> JInt(runtime.availableProcessors()),
+        "free_memory" -> JInt(runtime.freeMemory()),
```

Comment:
free memory will change over time, but this seems to get fixed the first time we upload something?

Reply:
I'm not sure this will be particularly helpful either way since it's the memory after the pipeline has been run.

Reply:
Fair, I can delete this.